hummingbot/connector/exchange/k2/k2_exchange.py

```diff
@@ -0,0 +1,858 @@
+
+        if tracked_order.is_cancelled:
+            self.logger().info(f"Successfully cancelled order {client_order_id}.")
+            self.trigger_event(MarketEvent.OrderCancelled,
+                               OrderCancelledEvent(
+                                   self.current_timestamp,
+                                   client_order_id))
+            tracked_order.cancelled_event.set()
+            self.stop_tracking_order(client_order_id)
+        elif tracked_order.is_failure:
+            self.logger().info(f"The order {client_order_id} has failed according to order status API. ")
+            self.trigger_event(MarketEvent.OrderFailure,
+                               MarketOrderFailureEvent(
+                                   self.current_timestamp,
+                                   client_order_id,
+                                   tracked_order.order_type
+                               ))
+            self.stop_tracking_order(client_order_id)
+
+    async def _process_trade_message(self, trade_msg: Dict[str, Any]):
+        """
+        Updates in-flight order and trigger order filled event for trade message received. Triggers order completed
+        event if the total executed amount equals to the specified order amount.
+        """
+        ex_order_id = trade_msg["orderid"]
+
+        client_order_id = None
+        for tracked_order in self.in_flight_orders.values():
+            if tracked_order.exchange_order_id == ex_order_id:
+                client_order_id = tracked_order.client_order_id
+                break
+
+        if client_order_id not in self.in_flight_orders:
+            return
+
+        tracked_order: K2InFlightOrder = self.in_flight_orders[client_order_id]
+
+        last_executed_amount: Decimal = tracked_order.executed_amount_base
+
+        updated = tracked_order.update_with_trade_update(trade_msg)
+        if not updated:
+            return
+
+        current_executed_amount: Decimal = tracked_order.executed_amount_base - last_executed_amount
+
+        fee_currency: str = ""
+        base_asset, quote_asset = tuple(tracked_order.trading_pair.split("-"))
+        fee_currency = base_asset if tracked_order.order_type == TradeType.BUY else quote_asset
+
+        self.trigger_event(
+            MarketEvent.OrderFilled,
+            OrderFilledEvent(
+                self.current_timestamp,
+                tracked_order.client_order_id,
+                tracked_order.trading_pair,
+                tracked_order.trade_type,
+                tracked_order.order_type,
+                Decimal(str(trade_msg["price"])),
+                current_executed_amount,
+                TradeFee(0.0, [(fee_currency, Decimal(str(trade_msg["fee"])))]),
+                exchange_trade_id=trade_msg["orderid"]
+            )
+        )
+        if math.isclose(tracked_order.executed_amount_base, tracked_order.amount) or \
+                tracked_order.executed_amount_base >= tracked_order.amount:
```

Comment:
Why would this happen? Is it ok to check for order filling more quantity than expected?

Reply:
Initially this was a rounding up issue.

But it could happen in the rare case when:
(1) Trade for a partial fill gets processed
(2) Websocket reconnects and subscribes to `SubscribeMyTrades` channel
- First response when subscribing to the channel would retrieve all recent trades.
- This would mean that the bot might reprocesses partial fills